### DISCUSSION TITLE
Best practice in changing logged-in user password

### DISCUSSION THREAD

**Discussion by bgeneto** [codeigniter4/shield#688] [Q&A] [answered]
Where in the documentation can I find the preferred method for changing a logged-in user's password?
Could someone plz show me the simplest way to achieve this task? 

If I use `$users->save()` or `update()` from user's Entity then it seems that I will need the users' email (even if I'm using username instead of email to login, i.e., commented email in the $validField config). And since the auth()->user() has no email property avail I need  to get the complete entity/model via `findById()`, even though the `user_id()` is known. 

I could use `setPassword()` but it seems to [have no effect](https://github.com/codeigniter4/shield/discussions/280). We need to call `saveEmailIdentity()` right after, even though I'm not using email for login (this name seems awkward to me). I think this need to be properly documented. 

Is there any straightforward (standard/preferred) way to change logged-in user's password using CI4 Shield? 

TIA.

**Comment by datamweb** [comment, marked as answer]
Please see [editing-a-user](https://codeigniter4.github.io/shield/quickstart/#editing-a-user).
An implementation is described in #639, which you can see in **step 5**.
If you think there is a need for a special explanation to change the password, you can send PRs.

**Reply by bgeneto** [reply]
Thanks for the great example in #639. As a suggestion, instead of using `UserModel` directly:
```
$users = model('UserModel');
```

I think it is better to use:

```
$users = model(setting('Auth.userProvider'));
```

And redirect with `forcePasswordResetRedirect()` instead of the static name `change-password`:

```
return redirect()->to(config('Auth')->forcePasswordResetRedirect())
```

I really think this example (improved as above) must be in the documentation somewhere.

**Reply by kenjis** [reply]
If you think so, why don't you send a PR?

**Reply by datamweb** [reply]
@bgeneto Thanks for your nice review. I updated #639 .

**Reply by kenjis** [reply]
`model(setting('Auth.userProvider'))` is correct, but you cannot autocomplete with `$users`.
`auth()->getProvider()` is better?

**Reply by datamweb** [reply]
@kenjis, do you mean the code below doesn't work?
```php
            // Success!
            $users = model(setting('Auth.userProvider'));
            $user = auth()->user()->fill([
                'password' => 'newPass'
            ]);
            $users->save($user);
```
I tested the above code, it seems to work fine.
Maybe I didn't understand what you mean.

**Reply by kenjis** [reply]
It does work.
I mean autocompletion in IDE. When you type `$users->s`, your IDE autocompletes `$users->save()`?

**Reply by datamweb** [reply]
Hmm, yes `auth()->getProvider()` looks better.

![demo1](https://user-images.githubusercontent.com/9530214/227137216-3a32a06d-b278-446e-8b1d-63a71fe91c4b.png)

![demo2](https://user-images.githubusercontent.com/9530214/227137246-7ebd5066-819c-4670-9965-f592b446aacd.png)
